### ISSUE TITLE
non-free dependency: Google AR Core

### ISSUE THREAD

**Issue by IzzySoft** [streetcomplete/StreetComplete#4289]
I've just scanned the APK built by F-Droid, and see:

```
Offending libs:
---------------
* Google ARCode (/com/google/ar/core): NonFreeDep

1 offenders.
```

This would mean all affected builds need to be disabled, as that's a no-go for F-Droid. So the questions are:

* I've looked at the corresponding smali code (generated via `apktool d de.westnordost.streetcomplete_4504.apk`), this does not look like a "stub". Am I mistaken?
* If it's no stub (which is to be assumed, as it's [explicitly implemented](https://github.com/streetcomplete/StreetComplete/blob/master/app/build.gradle.kts#L202)), is this essential – or can it be excluded e.g. by using a build flavor, or replaced by something else altogether?

Being a StreetComplete user myself, and just having made a big [announcement at Mastodon](https://mastodon.technology/@IzzyOnDroid/108820687843825531) how useful the app is, I'd hate seeing it disabled at F-Droid – so I really, really hope you can find a way to mitigate this.

**Comment by matkoniecz** [comment]
F-Droid can maintain a patch disabling this and remove functionality in few quest relying on it which measure width of some objects.

There is already fallback to measuring width with a tape and entering data manually.

**Comment by matkoniecz** [comment]
If SC can be changed without making it worse in way that makes maintaining patch easier, I guess that they can submit such pull request.

Or if there is a drop-in replacement for this library (afaik no) that also would be nice to know.

**Comment by licaon-kter** [comment]
I've seen and asked about this in https://github.com/streetcomplete/StreetComplete/pull/3709 at that time and was at ease since it said "but it's foss"

But now looking at https://github.com/google-ar/arcore-android-sdk and comparing with the AAR served by maven https://mvnrepository.com/artifact/com.google.ar/core/1.32.0 which has .so's and classes and whatnot, I fail to see WHERE is the source code exactly. :(

**Comment by IzzySoft** [comment]
Thanks for the fast answer! Indeed I'm not aware of any "smooth replacement" either (but I'm no dev, so I might simply have not heard of it). Thanks for the detail that it's at least not "crucially essential"! @licaon-kter maybe you could take a look at how to achieve that, so builds can be re-enabled?

**Comment by HolgerJeromin** [comment]
Adding some references for open AR replacements:
https://github.com/streetcomplete/StreetComplete/issues/2093#issuecomment-1090606995
https://github.com/microg/GmsCore/issues/1672

**Comment by westnordost** [comment]
**@IzzySoft so which consequences does this scan have at all?**

I'd not like to push an F-Droid version of the app that has the AR feature disabled. Only because one installed the app from F-Droid shouldn't mean that the choice is taken away from him whether he wants to install another non-free app to enable the AR feature or not. [As HolgerJeromin linked in the previous comment](https://github.com/streetcomplete/StreetComplete/issues/4289#issuecomment-1216733709), you do not even need Google Play Services as you may be able to sideload the ARCore app from another source.

**To clarify:**

1. The dependency in question is `com.google.ar:core` and that dependency itself **is** open source, see https://github.com/google-ar/arcore-android-sdk. So maybe that warning is imprecise(?)
2. That library does little else than to communicate with a separate non-free app on the smartphone called [Google Play Services for AR](https://play.google.com/store/apps/details?id=com.google.ar.core&hl=en&gl=US) (renamed from ARCore) which does the actual work. If it is not installed and Google Play Services are installed, the above library will ask the user to install it. If not, nothing happens and the AR feature simply does not work. (If I remember correctly, not even the button to start the AR measurement will be shown.)

**Comment by licaon-kter** [comment]
@westnordost can you read my comment above pls?

**Comment by westnordost** [comment]
> I fail to see WHERE is the source code exactly. :(

It is on github. Certainly not in the AAR, I do not think an AAR is meant to contain source code.

**Comment by westnordost** [comment]
Hmm right... there is not a lot in that repository, most of all, no java code.

**Comment by westnordost** [comment]
On the other hand, `com.google.ar:core` is listed as licensed under `Apache License, Version 2.0` on Maven Repository because in the POM file it is stated to be that license by the library author (Google LLC). So either Google violates their own terms in not also publishing the source code or they have [hidden it really well](https://hitchhikersguidequotes.tumblr.com/post/14333727462/mr-prosser-but-mr-dent-the-plans-have-been).

But I feel there is some missing information here. That 
```
* Google ARCode (/com/google/ar/core): NonFreeDep
```
is flagged can only come from that someone added that rule manually in some tool, right? So, how come it was added? Maybe a rationale can be found in the commit message (in the repo for whatever tool is this).

Anyway, at the moment I am not inclined to create a special version for F-Droid without this dependency. You'd have to do this yourself, albeit it is not a lot of work to create a patch.

The reason is that I don't accept that the (*alleged*) failure to comply with their own open source license of some dependency this app is using should lead to expulsion of this app from the F-Droid. As far as I am concerned, for me it's enough to know that the library is open source on paper, because that means that the app including all its dependencies may legally be distributed under these terms, i.e. including decompilation and modification of the code in e.g. the `com.google.ar:core` dependency.

**Comment by westnordost** [comment]
Nevermind my statement about the POM, I noticed that while it **says** Apache License, Version 2.0 in the name, it actually links to the LICENSE file in the repo. However, in that LICENSE file in the repo, they make clear that the additional terms of service only apply to 
- tools/arcoreimg/linux/arcoreimg
- tools/arcoreimg/macos/arcoreimg
- tools/arcoreimg/windows/arcoreimg.exe
- all files ending with *.apk

, with the rest being Apache License, Version 2.0. The AAR published via maven does not contain any of these.

**Comment by licaon-kter** [comment]
It's all fun and games until I ask again about the source? I've got nothing to build...

**Comment by westnordost** [comment]
How did it work before up until v45.2 then?

**Comment by licaon-kter** [comment]
How is it that we didn't scan for this dependency before? 

As usual, not enough hours in a day for contributors to port `nonfreedep` from one source ( https://gitlab.com/IzzyOnDroid/repo/-/tree/master/lib ) say via `grep -i nonfreedep libinfo.jsonl|cut -d ":" -f2|cut -d "\"" -f2|cut -d "/" -f2-|sed 's!/!.!g'` to the fdroidserver scanner ( https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/scanner.py#L62 ).

Blacklisting bad stuff is a never ending war, there's always one more.

**Comment by licaon-kter** [comment]
/rant @eighthave maybe merge some of the "scanner" MRs that await too? https://gitlab.com/fdroid/fdroidserver/-/merge_requests

**Comment by westnordost** [comment]
That was not my question. How did it build before if allegedly the F-Droid build builds each dependency itself from source instead of just the app from?

**Comment by licaon-kter** [comment]
We don't build everything from source, see: https://f-droid.org/en/docs/Inclusion_Policy/ (starting with _Though we tried to build everyting from source_).

As it would be a wasted effort when things are build already from source. That's the issue here, this dep is hosted on a ["trusted" maven repo](https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/scanner.py#L245) but they didn't do due dilligence to verify if truly FOSS. Not an isolated case. More info in this future post: https://gitlab.com/fdroid/fdroid-website/-/merge_requests/838/diffs#a517ae8bcf7835a17668dab09c873788052530af_0_5

**Comment by IzzySoft** [comment]
@westnordost I'd gladly state that I was wrong all along, and apologize in any way – if only the source were there. Until now, all we have is a binary AAR file and a statement that it's supposed to be Apache-2.0 – a license that states we "may reproduce and distribute copies of the Work or Derivative Works thereof in any medium, with or without modifications, and in Source or Object form" – thus granting us the freedoms required to be counted as "F/LOSS". But we cannot make use of that freedom if the source is not even there. There's not even the freedom to investigate the source, making sure what it does and that it corresponds to the AAR – if the source cannot be found.

All I see in [google-ar/arcore-android-sdk](https://github.com/google-ar/arcore-android-sdk) are some media files, C headers – and example code. But not the source code of this AAR. You've seen that for yourself. So if the source is not available, it cannot be "open source" – and thus cannot be accepted by F-Droid :cry: And no, I don't see any "alternatives" behind the links @HolgerJeromin posted: if I didn't miss something essential, they basically state that you can use AR with StreetComplete if you simply sideload that APK. Which is only possible because StreetComplete includes this AAR.

I fully understand your frustration, Tobias – and believe me or not, I fully share it. I just started using SC a little more than a month ago, and already I am so fond of it that I recommend it everywhere, motivating people to use it. I even made something I didn't do since my repo exists: I set up to have it [serve the expert edition](https://apt.izzysoft.de/fdroid/index/apk/de.westnordost.streetcomplete.expert), despite of its being more than two times over the size limit (I've never let an app enter my repo that way – that's rather the time I unlist an app). So I very much hope we can find a solution here. Ignoring the fact doesn't count as such (we cannot do that) – so it's either we turn up the source, or we need to "outsource" it somehow.

Now I thought about asking them at their repo – they won't accept that: all questions shall be asked at SO using the `ARCore` tag. [Searching the tag along with the keyword "source"](https://stackoverflow.com/search?q=%5Barcore%5D+source+answers%3A1&searchOn=3), I've stumbled upon [Is Sceneform 1.17.1 open source?](https://stackoverflow.com/questions/65332042/is-sceneform-1-17-1-open-source) – which states:

> As it is identical to 1.15.0 which is closed source, and the source is not provided separately, it is probably safest to assume it is actually closed

Looking into the smali of SC, there is `/smali/com/google/ar/sceneform` – so [where](https://github.com/streetcomplete/StreetComplete/blob/master/app/build.gradle.kts#L204) does this come from? AAR from the same source? For this at least there seems to be an open source variant available: [sceneform-android](https://github.com/SceneView/sceneform-android), but it unfortunately seems to depend on the very same ARCode we cannot find the source for (though it seems to be optional if you only need Sceneform itself). I could not find an answer to the sources of Core, though, nor did a web search succeed (just found the same repo, Google's docs, and plenty of how-to-use).

So if we cannot find the source, and neither a FOSS replacement: can that maybe moved to an addon entirely, so the app itself stays FOSS and thus at F-Droid?

**Comment by IzzySoft** [comment]
PS: For what it's worth, I started a [shout at Mastodon](https://mastodon.technology/@IzzyOnDroid/108834421860713820). Let's see if it turns something up…

**Comment by westnordost** [comment]
Sceneform is a separate library and is not a problem. The code is available.

---

@licaon-kter: so that makes this a question of policy rather than a question of necessity. You do not want to have any apps in F-Droid whose dependencies you could not **theoretically** build from source regardless of whether they are licensed with an open source license or not. 

I've read through the Apache license 2.0 and it looks like indeed the licensor does not commit himself to also publish the source, there is no mention of it. Hence, to my understanding, a work can be licensed with Apache 2.0 without the source to be available.

However,  this does not seem to touch on the rights of the licensee to produce derivative work, in other words,  decompile it,  make modifications, redistribute under the same terms:

> Redistribution. You may reproduce and distribute copies of the Work or Derivative Works thereof in any medium, with or without modifications, and in Source or Object form, provided that You meet the following conditions: 

<details>
<summary>(comment one condition)</summary>

> [...]
> (c) You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works; [...]

But since there is no (separate) source form of this work, this doesn't apply. Definition of source:

> "Source" form shall mean the preferred form for making modifications, including but not limited to software source code, documentation source, and configuration files.

What I understand from that is that as far as the license is concerned, the compiled blob of bytecode is the "preferred form for (providing for) making modifications" by the licensor, as there is no alternative. E.g. compare to a similarily licensed PNG file of e.g. some complex logo - in the absence of the GIMP "source" file,  the PNG is the preferred because only way to make modifications by the licensor.

Most importantly, it doesn't take away any right of the licensee to produce (proper) source code as a derivative work.
</details>

My point being: the licensee has every right to decompile the bytecode, make modifications and redistribute it, and that's what free software is about,  that you are granted the **right** to do it.  It is not about the **ease** of doing so. (An obvious flaw in non-GPL licenses)

Hence, I say that this a question of policy, i.e you do not want to include any apps whose dependencies do not also include the (human readable) source code in F-Droid. Regardless of the license.

---

Given that the apparent non-inclusion of the source code in the Apache licensed work was even a deliberate act rather an oversight on the basis of "meh, don't care" by Google ARCode devs, wouldn't it solve your policy issue if a decompiled version of that lib was available somewhere as a "derivative work"?

**Comment by mnalis** [comment]
> I've read through the Apache license 2.0 and it looks like indeed the licensor does not commit himself to also publish the source, there is no mention of it. Hence, to my understanding, a work can be licensed with Apache 2.0 without the source to be available.

Yes, it can. It is then however **_neither_** [free software](https://www.gnu.org/philosophy/free-sw.html) (i.e. lacks _"preferred form of the program for making changes in"_) **_nor_** [open source](https://opensource.com/resources/what-open-source) (i.e. is not _"software with source code that anyone can inspect, modify, and enhance."_) - license notwithstanding (kind of like Microsoft Windows is not free software nor open source software - although it contains, in binary form, significant amount of [BSD-licensed](https://en.wikipedia.org/wiki/BSD_licenses) code, and Microsoft Windows do show you all those licenses) 

[F-droid](https://f-droid.org/en/about/) is a repository of _"free and open source software (FOSS)"_. If software is not FOSS, it cannot (should not) be included there. And yes, that is because of their FOSS-only **policy** (which I personally happen to wholeheartedly approve of), which is arguably only reason for it's existence.

>>  _"Source" form shall mean the preferred form for making modifications, including but not limited to software source code, documentation source, and configuration files._

> What I understand from that is that as far as the license is concerned, the compiled blob of bytecode is the "preferred form for (providing for) making modifications" by the licensor, as there is no alternative

No, I'm pretty confident you misunderstood there. **_Source_** is _"preferred form for making modifications"_ -- not _"only form actually available publicly"_ or anything like that. So unless the original authors of software made that `.aar` library by typing numbers in hex editor, but **_preferred_** some other way of creating it, than that other way is actual **_preferred_** form of making modifications, and thus the Source is that what original authors used. Even [minified javascript](https://opensource.stackexchange.com/questions/11490/minified-gpl-code-inside-javascript-webapp) (for example) is **not** source code - because it is not the **_preferred_** way to write javascript.

> My point being: the licensee has every right to decompile the bytecode, make modifications and redistribute it, and that's what _free software_ is about, that you are granted the right to do it. It is not about the ease of doing so.

Well, it's **not** what [free software](https://www.gnu.org/philosophy/free-sw.html) is about, at least if you ask Free Software Foundation (/ RMS) who coined that term **"free software"**. See link above, specifically part starting with `Therefore, accessibility of source code is a necessary condition for free software. Obfuscated “source code” is not real source code and does not count as source code.`

> Hence, I say that this a question of policy, i.e you do not want to include any apps whose dependencies do not also include the (human readable) source code in F-Droid. Regardless of the license.

Yes, free software / open source / FOSS / f-droid is always about policy, regardless of license. In fact, you can have code which does not have _[any license at all](https://en.wikipedia.org/wiki/Software_license)_ (for example, because it is [Public Domain](https://en.wikipedia.org/wiki/Public_domain#Government_works), i.e. not covered by Copyright and related rights, and thus not needing license to relinquish that copyrights) - and it can still be free software and open source and included in f-droid.

---

> Assuming that the apparent non-inclusion of the source code in the Apache licensed work was even a deliberate act rather than an oversight on the basis of "meh, don't care" by Google ARCode devs, wouldn't it solve the policy issue if a decompiled version of that lib was available somewhere as a "derivative work"?

Uhhh, I wouldn't really want to venture there, nor have people I've become fond of paint such targets on their back for google lawyer-army to practice on, even if they can prove beyond doubt that that binary package indeed corresponds **in full** to that abandoned github page, and that it's indeed intended to be its license (which is far from conclusive).

**Comment by matkoniecz** [comment]
> in fact, you can have code which does not have [any license at all](https://en.wikipedia.org/wiki/Software_license) (for example, because it is [Public Domain](https://en.wikipedia.org/wiki/Public_domain#Government_works), i.e. not covered by Copyright and related rights

note that in this case it is better described as being licensed as public domain, not as having no license at all

in case of not having any license at all and not having copyright waived by law - works are fully copyrighted

**Comment by matkoniecz** [comment]
> Looking into the smali of SC, there is /smali/com/google/ar/sceneform – so [where](https://github.com/streetcomplete/StreetComplete/blob/master/app/build.gradle.kts#L204) does this come from?

I am confused by https://github.com/google-ar/arcore-android-sdk/issues/1049 - is sceneform being

- the same as one being published as open source by Google at some point and in new versions of arcore it is missing, because it is not developed and ended on Google graveyard
- developed further but as closed source available only as binary blobs (on which license? is it conflicting with GPL?)

**Comment by IzzySoft** [comment]
@matkoniecz If I read that correctly, Sceneform was abandoned. The last thing they did was fully open-source it¹, then archive its repo. There's a sill-alive and community-maintained [Sceneform fork](https://github.com/SceneView/sceneform-android) available, though.

¹ <sub>v.15 was closed source, 1.16 open-source with parts missing, v1.17 then added missing source parts</sub>

My toot on Mastodon meanwhile brought in some answers. Basically, nobody seems to be able to find the full source; even Marvin W (author of microG) states [it is incomplete](https://github.com/SceneView/sceneform-android). Another dev will raise the question in Google's issue tracker today; let's hope we get a positive answer there.

**Comment by matkoniecz** [comment]
> v1.17 then added missing source parts (...) nobody seems to be able to find the full source

So StreetComplete is using an abandoned project that was open sourced - but in way that left part of code not actually published?

In such case I would at least try asking to publish also missing parts. So "another dev will raise the question in Google's issue tracker today" seems a good next step.

**Comment by IzzySoft** [comment]
Here's the link to the issue: https://issuetracker.google.com/issues/242730604 (whoever has a Google account can give it a +1 or chime in).

**Comment by matkoniecz** [comment]
> https://issuetracker.google.com/issues/242730604

I guess that it would be helpful to point out specific missing parts, if any were identified to be missing.

> even Marvin W (author of microG) states [it is incomplete](https://github.com/SceneView/sceneform-android)

ling goes to some general repo, not to statement confirming this

**Comment by westnordost** [comment]
Hey guys,I found the sources:

**https://dl.google.com/dl/android/maven2/com/google/ar/core/1.32.0/core-1.32.0-sources.jar**

When clicking the "View all" link on https://mvnrepository.com/artifact/com.google.ar/core/1.32.0 , the Google repo returns 404 on the directory listing. However, the naming scheme is always the same in maven, so I guessed the name.

**Comment by licaon-kter** [comment]
The top message of the .java files is fun:
```
/*
 * Copyright 2018 Google LLC
 * CONFIDENTIAL. AUTHORIZED USE ONLY. DO NOT REDISTRIBUTE.
 */
```

**Comment by westnordost** [comment]
```
(╯°□°）╯︵ ┻━┻ 
```

**Comment by westnordost** [comment]
https://github.com/google-ar/arcore-android-sdk/issues/1538

**Comment by westnordost** [comment]
Okay, the dev(s) just confirmed even though the Java stuff is actually Apache licensed 2.0 and the above mentioned copyright notice is a "bug", there is a binary blob within that AAR which is proprietary, making the library as a whole not open source. 

So, AR functionality must be removed for F-Droid publication.

**Comment by matkoniecz** [comment]
But it still can be used in SC itself, right? Is https://developers.google.com/ar/develop/terms conflicting with GPL for SC code in general?

**Comment by westnordost** [comment]
FYI I plan to defer working on this to re-enable F-Droid builds because I like to first wait a bit to see if the Google folks at arcore-android-sdk have a change of heart. It doesn't seem to be a big team, maybe they can make such decisions on their own measure.

If anyone wants to see F-Droid builds pop up sooner again, you are welcome to create a PR. I am not sure how custom builds for F-Droid should be handled.
Of course the PR must not remove the AR stuff, but just make it so that (a newly to be defined) build flavor for F-Droid does not include the AR-specific sources and dependencies. I am not sure how to do that, you'd need to read some gradle documentation. Or alternatively, maybe F-Droid folks offer another mechanism how to patch an app for the F-Droid build bot: Maybe it is enough to write a "patch" that automatically gets applied before each build? In either case, it necessitates some reading into documentation.

**Comment by IzzySoft** [comment]
@westnordost as re-establishing SC in F-Droid might take a bit then (@licaon-kter just meant to say the option exists, not that a patch is in the works) – if you think it's needed and would help, I'd make a second exception from the size-rule and offer SC an "interim home" in my repo until it returns to F-Droid.org. Signatures would not match, so it won't help existing F-Droid users (myself included) much unless they are ready to uninstall and reinstall and it might be easier for them staying with the installed version or (temporarily) use the extended edition (available via my repo) alongside – but the offer stands.

Speaking of the EE: can existing data from SC be transferred to SC EE (and back) in some way?

**Comment by westnordost** [comment]
Nah, I do not think this would make sense. People can just use either the version from Google Play or from GitHub if they want the newest features. For EE, you'd need to ask @Helium314 if he changed any database tables or names in the shared preferences. If not, you can move the directories where the data lies accordingly, but you need to have full file system access to access the private folders of the apps (root, maybe adb?).

**Comment by IzzySoft** [comment]
Thanks! I thought I might have missed an export/import feature (would be a different issue then, if you'd accept that feature request, as it'd help "migration").

**Comment by matkoniecz** [comment]
for import/export see https://github.com/streetcomplete/StreetComplete/issues/3481

**Comment by Helium314** [comment]
Database and preferences files can be transferred between EE and vanilla SC, only the `TangramPinsSpriteSheet` preferences should be removed from preferences file, or quest pins/icons may be missing or messed up.

**Comment by mnalis** [comment]
> But it still can be used in SC itself, right? 

**No**, not at all - i.e. not unless:
- the missing source code is released (under some [GPLv3-compatible license](https://www.gnu.org/licenses/license-list.html#GPLCompatibleLicenses)),  **_or_**
- **all** StreetComplete contributors **_jointly agree_** to re-license under some (non-[copyleft](https://en.wikipedia.org/wiki/Copyleft)) license which is compatible with non-free (AKA closed-source / proprietary) software. 

`StreetComplete.apk` is supposed to be licensed under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html). Section 6 **_"Conveying Non-Source Forms"_** states (among other things) _"You may convey a covered work in object code form under the terms of sections 4 and 5, provided that you also convey the machine-readable Corresponding Source under the terms of this License"_

As the source of the part of the `.apk` is missing (no source for `ARcore` at the moment), we are physically unable to _"convey the machine-readable **Corresponding Source**_" in **any** way - and thus **_may not_** "_convey a covered work in object code_" -- i.e. may not distribute affected `.apk` files at all. 
We may continue to distribute the source from https://github.com/streetcomplete/StreetComplete/ (which is all [FOSS](https://en.wikipedia.org/wiki/Free_and_open-source_software)) and ask all users to compile `.apk` files themselves (as GPLv3 restrict  distribution only, not usage), but that would of course be highly inconvenient to 99% of the users (to say the least!)

So to continue distributing `.apk` files **anywhere** (GooglePlay, GitHub, F-Droid etc) under current license we need to solve this issue:

- One way it could be done is to remove all ARcore non-free software completely (i.e. lose automatic width/height measurements in those quests).
- Another would be to put **all** ARcode related code in separate (optional) companion `StreetComplete-nonfree-ARcore.apk` plugin package - like OsmAnd used to do in times past for its Countours and Parking plugins for example), and only call the functions there if the users installed that plugin. The users who would would to continue using AR functinality in quests would need to install that `plugin.apk` package in addition to main `SC.apk` package.
- or we get lucky soon (before new SC release) and find ARcore sources under GPL-compatible license

**Comment by FloEdelmann** [comment]
Is the compiled `StreetComplete.apk` really GPLv3-licensed? Or only the source code?

**Comment by westnordost** [comment]
Uff, I didn't know that. I was under the impression that GPL would only be viral in that if a GPL licensed piece of code is used (e.g. a library), the (binary) distribution as a whole must also be GPL, not the other way round, i.e. all code contained in the binary distribution must be GPL compatible too (e.g. not only the code owned by this project).

This means that it is not even an F-Droid issue anymore. That Google's ARCore library contains non open-source code means that **any** copyleft-licensed project cannot use that ARCore, distributed anywhere, which is not made transparent at all (on the contrary).

I figure this creates a lot of work or even makes the work invested so far moot for any copyleft licensed open source project that has been lured into using or even basing their software on Google's false statement that the ARCore Android library is open source. I do not think there was a deliberate intent but it could surely be interpreted as feigning false facts.

<strike>The non open-source object in the library has an estimate of 2000 lines of code (60KB, [see this paper](https://www.leshatton.org/Documents/LOC2005.pdf)) but this including any included dependencies as Abseil (, boost?, ...). So, given that the native part of that library likely only does the communication to the proprietary APK, it may be easy to reverse-engineer this. One may interpret that the license currently says that (at least) the object form of this native part is Apache licensed.</strike> However, I will not do this because 1. I noticed that not even the Java sources are complete. In the linked source JAR, many methods have their contents removed, 2. legal uncertainty regarding the reverse-engineering, 3. I want to wait on a whether they will consider open-sourcing the closed-source part of their library. 4. It is a lot of work

In any case, what I intend to do is:

1. Add another comment in the linked github ticket mentioning the fact that @mnalis described, maybe hinting at that it is impermissible to state that a work is (i.e. license it as) open source and later retract that by saying it was an oversight - once granted, a license cannot be revoked. Or, at least, this is feigning false facts, deliberate or not.
  In any case, I will prompt them to give an answer to whether they will release the source for that part or not, so we know how to continue
2. I will wait for their response whether they do intent / will consider to make this open source and meanwhile publish new versions of StreetComplete as usual
3. If eventually they respond that they do not intend / will not consider to make this open source, I will create a separate Apache-licensed app and publish it to the app store that contains all the code for the measuring (on the upside: People could then use that app to also measure things independently of StreetComplete). StreetComplete will then simply ask the user to download that app from the play store or github when it is not installed and the user taps on the "measure with AR" button

**Comment by IzzySoft** [comment]
Ugh, that becomes much heavier than I thought… So if case 3) should come true (which we all hope will not), how will you go for the F-Droid build? Reading between the lines ("a separate Apache-licensed app"), the current one will have ARCode stripped so we can build it again? Just asking for clarification. And thanks for all your efforts!!!

**Comment by matkoniecz** [comment]
The entire point of such complicated splitting would be to eliminate tainted code from StreetComplete.

Though depending on what F-droid needs: is it ok to propose to user to install proprietary app? If no then F-droid would need to patch it away

**Comment by licaon-kter** [comment]
`NonFreeAdd` [anti-feature](https://f-droid.org/en/docs/Anti-Features/#NonFreeAdd)?

**Comment by IzzySoft** [comment]
That would earn you the NonFreeAdd anti-feature, yes. But it would be acceptable, yes.

**Comment by westnordost** [comment]
Wouldn't it be the same already even if the Android ARCore library way open source? After all, it "promotes" (i.e. doesn't work without) the "Google Play Services for AR" APK and even offers a short-cut to install it AFAIK.

**Comment by IzzySoft** [comment]
Yes and no. The description says "promotes or entirely depends on". So if it's just a possibility (as it currently is – at least I wasn't yet pointed to ARCode in the app), it would not necessarily raise the AF. Border might be "thin" or "arguable" – but that's how I would categorize it here. When in doubt I usually consult with e.g. @licaon-kter for a second opinion :wink:

**Comment by licaon-kter** [comment]
@westnordost How would it work exactly? Pop-up? Option?

**Comment by mnalis** [comment]
> Is the compiled StreetComplete.apk really GPLv3-licensed?

Yes. That is the nature of [copyleft](https://en.wikipedia.org/wiki/Copyleft) / [viral licenses](https://en.wikipedia.org/wiki/Viral_license) like GPL3 or CC-BY-**SA**. Linked wikipedia articles explain it in layman terms as _"[...] which requires that any derivative work also be licensed under compatible licenses with the GPL"_. The actual term to search for in GPLv3 legalese is _"covered works"_ (and act of <em>convey</em>ing it).

> Uff, I didn't know that. I was under the impression that GPL would only be viral in that if a GPL licensed piece of code is used (e.g. a library), the (binary) distribution as a whole must also be GPL, not the other way round (e.g. it is not enough if all the code owned by the project is GPL, also all dependencies must be compatible to GPL).

If you think it about it, it would be trivial to circumvent  GPL copyleft protections if it mattered which part is a library and which is a main program (e.g. someone wishing to add GPL library in their non-free program, could simply make their proprietary program a library instead, and then make trivial GPL "main program" which consists only of calling both free and non-free libraries).  So strong copyleft licenses like **GPL prohibit distributing a mix a free software (open source) with non-free software (closed source)** - no matter the amount of free or non-free code or their location ("main program" or "library").

There are other class of free software licenses (jointly called [permissive licenses](https://en.wikipedia.org/wiki/Permissive_software_license)) like mentioned [Apache license](https://en.wikipedia.org/wiki/Apache_License) which do not guarantee [copyleft](https://en.wikipedia.org/wiki/Copyleft) protections (i.e. while the code itself is free/open source it may be mixed with non-free code are resulting binary may be non-free/closed-source, including a non-free fork of library like famous [Embrace,Extend,Extinguish](https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish) tactics). And then there are licenses like [LGPL](https://en.wikipedia.org/wiki/GNU_Lesser_General_Public_License) which are in between. 

So, to simplify:
- "source GPL" + _anything_ => result must be GPL (so "_anything_" must be compatible with GPL, or you may not distribute the result at all)
- "source Apache" + _anything_ => result may be _anything_

**Comment by westnordost** [comment]
@licaon-kter Amongst many other things, the app may ask about what is the width of some road. When answering this, an input field to input the width is displayed. Next to this input field, there is a "measure with AR" button. Clicking this button will either open the AR measurement view if ARCore is installed or ask to install it if it is not. After AR measurement is complete, whatever was the result is pre-filled into the input field.

**Comment by westnordost** [comment]
@mnalis are there any licenses that enforce a "share alike" but only for the code owned by that project?

**Comment by mnalis** [comment]
> @mnalis are there any licenses that enforce a "share alike" but only for the code owned by that project?

Yes, the [Lesser GPL (LGPL)](https://en.wikipedia.org/wiki/GNU_Lesser_General_Public_License) that I just mentioned. But it depends on what exactly you want to use it for (and how tolerable one is to losing some of GPL protections). It is mostly used only for libraries, but there are exceptions IIRC.

**Comment by mnalis** [comment]
> Wouldn't it be the same already even if the Android ARCore library was open source? After all, it "promotes" (i.e. doesn't work without) the "Google Play Services for AR" APK and even offers a short-cut to install it AFAIK.

[Non-Free Addons](https://f-droid.org/en/docs/Anti-Features/#NonFreeAdd) antifeature says _"This Anti-Feature is applied to apps that, although Free Software themselves, promote other Non-Free applications or plugins."_ 
While it is f-droid discretion to interpret it, it would look to me:

- if the **app** itself **_does not_** explicitely suggest to install non-free add-on (i.e. it's not saying _"click here to install this addon for extra features"_), then IMHO it does not deserve antifeature, even if non-free addons are available on Google Play or mentioned online in FAQ etc. That would be the case if the app silently auto-detected if add-on is available and use it if it exists, or don't use it it doesn't (that's how OsmAnd Parking plugin used to work, for example)
- however, if the **app** itself explicitly suggest installing non-free add-on (like in _"Clicking this button will either open the AR measurement view if ARCore is installed or **ask to install it** if it is not"_), then it seems likely to me to get marked with [Non-Free Addons](https://f-droid.org/en/docs/Anti-Features/#NonFreeAdd). That also means that StreetComplete would not be even shown to F-Droid users which opted to not want apps with that antifeature.

**Comment by mnalis** [comment]
Also, according to https://github.com/google-ar/arcore-android-sdk/blob/master/README.md it seems that _"By downloading the ARCore SDK for Android, you agree that the [ARCore Additional Terms of Service](https://developers.google.com/ar/develop/terms) governs your use thereof."_

And those seems to require (among other things - as there are several sub-documents linked in which I have not yet looked at!) updating privacy policy (of the new addon _or_ the main SC app, depending how the situation turns out) with extra cruft like:

> The terms of service for your API Client (e.g., your app) will (1) notify end users that the API Client includes ARCore functionality provided by Google; and (2) state that use of ARCore functionality is subject to the then-current versions of the: (A) Google Terms of Service at [policies.google.com/terms](https://policies.google.com/terms); and (B) Google Privacy Policy at [google.com/policies/privacy](https://google.com/policies/privacy).

I've <strike>starting to think</strike> becoming convinced that even if they do decide to release the full source, it would still be prudent to at least remove all ARcore functionality to separate SC addon (instead of forcing all that google baggage on all unsuspecting SC users - i.e. if it remains in main app we can _**kiss goodbye**_ that nice SC privacy preamble _"Ah, someone who cares about privacy, nice! I think I have only good news for you..."_  :cry: )

**Comment by mnalis** [comment]
> FYI I plan to defer working on this to re-enable F-Droid builds because I like to first wait a bit to see if the Google folks at arcore-android-sdk have a change of heart. It doesn't seem to be a big team, maybe they can make such decisions on their own measure.

It has been a month since it [has been clarified](https://github.com/google-ar/arcore-android-sdk/issues/1538#issuecomment-1217888964) that ARcore library is indeed incompatible with GPL, with no further comments from Google side since then. It thus seems unlikely to me that _"maybe they can make such decisions on their own measure"_ and we should probably start considering how to address that for future releases, i.e.:

> 3. If eventually they respond that they do not intend / will not consider to make this open source, I will create a separate Apache-licensed app and publish it to the app store that contains all the code for the measuring (on the upside: People could then use that app to also measure things independently of StreetComplete). StreetComplete will then simply ask the user to download that app from the play store or github when it is not installed and the user taps on the "measure with AR" button

However, making a separate app will likely take some time, and [f-droid is stuck](https://f-droid.org/en/packages/de.westnordost.streetcomplete/) on quite old `40.2`, which already confuses users (e.g. https://github.com/streetcomplete/StreetComplete/issues/4379). It would make sense to do a [patch that removes ARcore support from f-droid build](https://github.com/streetcomplete/StreetComplete/issues/4289#issuecomment-1216798809) so F-DROID can continue to get updated in the meantime. Would that be acceptable interim solution @westnordost? Or do you foresee better short term solution?

**Comment by westnordost** [comment]
As a short term solution, this would be acceptable. It should be done in one clean PR so it can be reverted once the full solution is done (which is either simply revert because Google made it properly open source or which is creating a new non-GPL app that does the measuring and report back to StreetComplete)

**Comment by mnalis** [comment]
I did a [remove-arcore-nonfree](https://github.com/mnalis/StreetComplete/tree/remove-arcore-nonfree) branch that removes `com.google.ar*` and related stuff based on latest `v47.0-beta1` (which happens to be `master`), and when testing  [it seems to work fine](https://github.com/mnalis/StreetComplete/actions/runs/3093924358), i.e. does not show AR icon and collects manually entered values normally. Hopefully I didn't forget anything.

@IzzySoft do you know how to correctly push that patch to f-droid so it can publish new version when it arrives?

**Comment by westnordost** [comment]
Are then the quests always disabled by default? The quests should be disabled by default if no AR measure is available (because you need to have a measuring tool with you otherwise to solve the quest)

**Comment by westnordost** [comment]
It would be fine to push this to a branch on StreetComplete, e.g. "no-arcore" too. However, not sure how the releases with F-Droid should happen then - currently it is done via tags (e.g. `v47.0-beta1`). But the tags will continue to be set on `master`, not on `no-arcore`. So what's the best solution here?

**Comment by TheLastProject** [comment]
Most Android developers tend to solve these things using "gradle flavors", so it's all in one codebase. F-Droid can build the fdroid/foss/whatever flavor then.

However, if you don't want to mess with gradle flavours you could maybe just each release:
1. Rebase/merge new code into no-arcode branch
2. Create a v47.0-beta1-fdroid tag (with the appropriate number)

(Personally, I would always re-base so the no-arcode patch is always the patch "on top" of master).

Then F-Droid can just change the metadata rules so it only picks up the -fdroid tags.

If that's all too inconvenient, F-Droid can also apply a patch during build. See https://gitlab.com/fdroid/fdroiddata/-/blob/bc5bc6c329f6b3d72bc79086ab028648018dec2c/metadata/im.vector.app.yml#L943-944 for example. In such a case the patch file updating would have to go through the F-Droid team as it's in fdroiddata then, so this is the slowest option (if it's not updated on time the build will fail and updates need to go through the F-Droid team).

In the end it all comes down to your preference I think, if you're only planning on removing arcore from StreetComplete for like one or two releases, the fdroiddata patch method might be easiest. But for long-term it can get quite frustrating.

**Comment by mnalis** [comment]
> Are then the quests always disabled by default? The quests should be disabled by default if no AR measure is available (because you need to have a measuring tool with you otherwise to solve the quest)

Yes, quests are disabled by default - the patch basically just makes `ArSupportChecker()` return `false` (and removes all offending imports and affected code), and so the SC then behaves as if the phone does not have AR support (so it disables quests by default and does not show AR button in them).

> So what's the best solution here?

Ah, I have no experience there, but as @TheLastProject notes, and since we plan to make AR functionality a plugin anyway in relatively near future (or even Google might decide to open source it), the f-droid patch solution would seem simplest to me. Thanks to @TheLastProject  example, I could make and submit such MR to fdroiddata, if all agree that is the best approach?

**Comment by TheLastProject** [comment]
They're regular git patches. Git can do some small resolving and generally doesn't mind if unrelated sections get changed (as you can see, I can re-use the Element patches for several versions), but it's really dependent on git itself.

In the patch no longer applies cleanly, git will throw an error and the build will fail until the patch is fixed by someone.

**Comment by westnordost** [comment]
I think then the solution to post the patch at F-Droid is the easier solution because of the relative isolation of the code and no necessity to rebase+add a fdroid specific tag to a branch on each release

**Comment by mnalis** [comment]
OK, I've done a fdroiddata merge request here: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/11785

@TheLastProject you seem to have experience with the process, if you have a minute, could you take a look and say if that looks good to you? I've tested that the patch applies against `v46.0`, `v46.1`, `v47.0-alpha1` and `v47.0-beta1`; also  CI/CD for that MR on GitLab seem to have passed OK. 

I've updated added a new block for `v46.0` with a patch, do you know will f-droid automatically realize it should also use that patch for `v46.1`, `v47.0` and newer when it tries to add them? Or what would be the procedure for the future versions?

**Comment by TheLastProject** [comment]
I see linsui responded to the MR already, they are a lot more active in fdroiddata than I am so I'd recommend listening to them over me :)

> I've updated added a new block for v46.0 with a patch, do you know will f-droid automatically realize it should also use that patch for v46.1, v47.0 and newer when it tries to add them? Or what would be the procedure for the future versions?

Well, right now update checking is disabled. But if all is good it can be re-enabled. With Auto update checking enabled, F-Droid copies the build instructions of the previous version so it'll automatically try the exact same patches for new releases too.

Talking about new releases, is there any reason v46.0 is being added to F-Droid instead of v46.1, given v46.1 seems to be the latest stable?

**Comment by mnalis** [comment]
> I see linsui responded to the MR already, they are a lot more active in fdroiddata than I am so I'd recommend listening to them over me :)

Yup, I'll have to do more research there where to put extra commnads. Your method of just providing patch seemed simpler to me than patch + rm + sed proposed in MR, but I can see how the later would be more robust...

> Well, right now update checking is disabled. But if all is good it can be re-enabled. With Auto update checking enabled, F-Droid copies the build instructions of the previous version so it'll automatically try the exact same patches for new releases too.

That is good news! Is that auto checking something I myself can enable in a patch (I see there is field called `AutoUpdateMode: None`, but it  has that value in your example too, so I suspect that is not it), or is it more of fdroiddata admin thing?

> Talking about new releases, is there any reason v46.0 is being added to F-Droid instead of v46.1, given v46.1 seems to be the latest stable?

well, `v46.0` was the first where the patch applied cleanly (it fails on `v45.x` and lower), so I though to go with that, and then see if autoupdate code picks it up for `v46.1` automatically.

**Comment by TheLastProject** [comment]
> That is good news! Is that auto checking something I myself can enable in a patch (I see there is field called AutoUpdateMode: None, but it has that value in your example too, so I suspect that is not it), or is it more of fdroiddata admin thing?

I've added a suggestion on the MR, let's see what linsui thinks :)

> well, v46.0 was the first where the patch applied cleanly (it fails on v45.x and lower), so I though to go with that, and then see if autoupdate code picks it up for v46.1 automatically.

I'm mostly wondering because it seems a bit weird to me to go this route unless there's a reason someone would want to explicitly run v46.0 and not v46.1.

Worst case scenario, the next build cycle starts after the merge but before the next checkupdates, which means users will first get an update to v46.0 and then a few days later an update to v46.1, which would mean users would first update to a version that has some known bugs already fixed in v46.1 and they'd be bothered with a new update again real quickly. Given each update is 56MB (looking at historic data), that could be a bit harsh on users with bandwidth caps with no real gain as far as I can tell.

**Comment by mnalis** [comment]
> I've added a suggestion on the MR, let's see what linsui thinks :)

thanks!

> I'm mostly wondering because it seems a bit weird to me to go this route unless there's a reason someone would want to explicitly run v46.0 and not v46.1.

People might want to try to pinpoint in what version something broke... But yeah, it is a little thin for a reason :smiley: 
So I've updated MR to reference `v46.1` now.

**Comment by IzzySoft** [comment]
@mnalis apologies I didn't respond to your ping in time (I was on vacation) – luckily Silvia (@TheLastProject) jumped in with the same track I'd have proposed. Glad a way was found to get SC updated at F-Droid again, looking forward to the update! I became quite addicted to the app I have to admit and am quite proud of my ranking :see_no_evil: Cyprus needed a lot of SC-love, and it received it during the past 2 weeks :star_struck:

**Comment by riQQ** [comment]
StreetComplete 46.1 without AR core is available on F-Droid as of yesterday, September 29th.
Thanks @mnalis

**Comment by opk12** [comment]
> licaon-kter: so that makes this a question of policy rather than a question of necessity. You do not want to have any apps in F-Droid whose dependencies you could not **theoretically** build from source regardless of whether they are licensed with an open source license or not.

No, the source is necessary to ensure that the binary is malware-free, by public scrutiny and reproducible builds (see https://f-droid.org/docs/Reproducible_Builds/ ): when independent builds give a byte-per-byte-identical result, you do not have to trust the build machine, kids, roommates and random guests of whoever submitted each lib binary. That these security hygiene concepts do not exist in Google Play, Windows and generally compulsory education / public legislation is a shame, not something that should be encouraged. This can't be done through decompiling, as it won't help with native code and R8-minified stuff, and is illegal unless allowed in the license.

**Comment by FloEdelmann** [comment]
Version 47.0 is now [available on F-Droid](https://www.f-droid.org/de/packages/de.westnordost.streetcomplete/), so everything is working as expected again. Thanks to everyone involved fixing this!

**Comment by IzzySoft** [comment]
Thanks a lot for all the efforts all of you have put into this. Very much appreciated, and makes me love (and recommend) the app even more! :heart_eyes:

**Comment by mnalis** [comment]
@FloEdelmann although, we should probably open another issue linked to this one - while F-droid builds now work, the issue with ARcore is **_not fixed_** - SC is still in copyright breach on Google Play (and [GitHub Releases page](https://github.com/streetcomplete/StreetComplete/releases)) and a final solution (either Google opensourcing the lib, _or_ the ARcode being moved to separate companion app + f-droid patch being removed) is still pending. 

See this comment for example:
https://github.com/streetcomplete/StreetComplete/issues/4289#issuecomment-1225586563

**Comment by westnordost** [comment]
The Google ARCore team seems to have finally done something about the open license issue by clarifying in the LICENSE.md that indeed the library is **not** open source (but only some related files in that repo are open source):

https://github.com/google-ar/arcore-android-sdk/commit/6f887e14366c5737621f91e3d8e29f520016e25d#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7R11-R13

So, with the certainty now here, it is time to remove ARCore from StreetComplete.

**Comment by mnalis** [comment]
Just a short note: before StreetComplete version with removed ARCore is released, we should also undo [f-droid MR 11785](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/11785), so F-droid builds can continue to build new version normally too (as it will take some time for new MR to be accepted there).

I can handle that f-droid part again (if you prefer), when the patch removing ARcore lands in StreetComplete master branch.

**Comment by westnordost** [comment]
Done now. For v51.0 onwards, the [[f-droid MR 11785](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/11785)](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/11785) can be undone

**Comment by IzzySoft** [comment]
Thanks Tobias! I've forwarded your message to the corresponding MR, asking my two colleagues who were working on it to take care (and linking them here).

**Comment by IzzySoft** [comment]
It's done.
